### PR TITLE
Overhaul `symbols-view`

### PR DESCRIPTION
**Pull request by savetheclocktower**
I should've submitted this earlier; I know it's asking a lot to have someone review this so close to 1.112. If it gets done, great; if not, it can wait a month.

### Issue or RFC Endorsed by Atom's Maintainers

#814

### Description of the Change

Changes `symbols-view` to be a package that follows the provider/consumer model, akin to `autocomplete-plus`. Instead of having its own symbol generation strategy, it exposes an interface and allows other packages to act as its data source.

The goal here is that an ordinary non-bleeding-edge user — someone who was very happy with their Atom experience and who hasn't enabled experimental Tree-sitter grammars — should not notice that anything is different. A Pulsar user who _has_ enabled experimental Tree-sitter grammars might notice that their symbol list is richer and more accurate. A Pulsar user who's using one of my experimental IDE packages may notice that the symbol list is _much_ richer — and that they can now do project-wide symbol searches and jump to the definition of the token underneath their cursor. In each case, we expect that the symbol navigation experience will be equal to or better than what they had before.

### Alternate Designs

Other options for how to bring this into core were mooted in #814. The team had a slight preference for this strategy.

### Possible Drawbacks

I’ve written unit tests and I can verify that this architecture works for me locally, but I have not done aggressive integration testing across multiple platforms. I am hoping that team members on Windows and Linux can let me know if anything obvious is broken.

### Verification Process

Ensure that you **are not** subscribed to `symbols-view-redux`, `symbol-provider-ctags`, or `symbol-provider-tree-sitter`. Check out this PR and relaunch Pulsar.

Here are some things to try in order to determine whether you’re using the new architecture:

* When you’re using a file without a Tree-sitter grammar, you shouldn’t notice that anything has changed when you invoke **Symbols View: Toggle File Symbols**. (LESS files and CoffeeScript files are two notable examples of Tree-sitter-less languages, so maybe try opening your user stylesheet.)
* When you’re using a file with a Tree-sitter grammar, the Tree-sitter symbol provider should be used. You’ll see evidence of it in the menu in the form of
  * “badges” for each symbol (function, class, etc.);
  * the ability to show symbols even in brand-new files that haven’t been saved to disk;
  * recursive naming behavior in certain file types (try opening a `package.json` and searching for a key name!).
* If you want to put it through an even more thorough workout, try installing [pulsar-ide-typescript-alpha](https://web.pulsar-edit.dev/packages/pulsar-ide-typescript-alpha), enabling the “Include JavaScript” setting, and reloading the window. This new package should be able to act as a symbol provider for both **Toggle File Symbols** and **Toggle Project Symbols**. In the latter case, you should be able to press <kbd>Cmd-Shift-R</kbd>/<kbd>Ctrl-Shift-R</kbd> and search for the name of a function or a class that exists anywhere in the project.

### Release Notes

* Enabled the core `symbols-view` package to accept symbols from a number of sources, including Tree-sitter grammars and IDE packages.

### CODE REVIEW

**Comment by savetheclocktower** [comment]
I think this is in a good place for review. Let's try to get it into the 1.113 release!

**Comment by Spiker985** [comment]
**Testing process**

1) Cloned branch into integrate folder: `git clone https://github.com/savetheclocktower/pulsar.git -b integrate-symbols-view-redux integrate`
2) Moved existing `symbol-provider-tree-sitter` package: `mv ~/.pulsar/packages/symbol-provider-tree-sitter ~/.pulsar`
3) Copied new symbol* folders from `integrate`: `cp -r ~/git/integrate/packages/symbol* ~/.pulsar/packages`
4) Installed required modules per package:
`cd ~/.pulsar/packages/symbol-provider-ctags; pulsar -p install`
`cd ~/.pulsar/packages/symbol-provider-tree-sitter pulsar -p install`
`cd ~/.pulsar/packages/symbols-view; pulsar -p install`
5) Open Pulsar instance
6) Edit -> Stylesheet
7) Ctrl + Shift + P -> Symbols View: Show Active Providers
image
8) Ctrl + Shift + P -> Symbols View: Toggle File Symbols
image
9) Opened Pulsar Source Code (src/main-process/win-shell.js) -> Symbols View: Toggle File Symbols
image
10) Opened a package.json -> Symbols View: Toggle File Symbols
image
11) Installed `pulsar-ide-typescript alpha`, Enabled `Include JavaScript` setting
12) Ctrl + Shift + P -> Pulsar IDE Typescript Alpha; Start Language Server
13) Ctrl + Shift + P -> Symbols View: Toggle Project Symbols -> "getApp"
image
14) Ctrl + Shift + P -> Symbols View: Show Active Providers
image

From the above testing, it appears that the packages are working as expected, however, I only worked on `.less`, `.js` and `.json` files predominantly.

I did test it out on a Rust project, and the `symbol-provider-ctags` provider was able to parse file symbols.

Conversely, I opened a PowerShell project (which doesn't have any kind of native ctag regex or support) and, as expected, was unable to see any symbols. I also generated a tags file with `ctags -R` within my project root and was unable to get the completion from that (however, I don't believe that to be a current feature).

From my experience, this integration appears to be successful and I approve of its merge into Core

**Review by Spiker985 — APPROVED** [review]

**Comment by savetheclocktower** [comment]
> I did test it out on a Rust project, and the symbol-provider-ctags provider was able to parse file symbols.

@Spiker985, are you sure those were coming from `symbol-provider-ctags`? I ask because [we've got a `tags.scm` for Rust](https://github.com/pulsar-edit/pulsar/blob/master/packages/language-rust-bundled/grammars/tree-sitter-rust/queries/tags.scm), so it would've been using that unless `symbol-provider-tree-sitter` had been disabled.

Just making sure we haven't got a bug we need to deal with at this late hour. If you're unsure which provider was supplying symbols, there's a setting to show the provider name in the list.

**Comment by Spiker985** [comment]
> > I did test it out on a Rust project, and the symbol-provider-ctags provider was able to parse file symbols.
> 
> @Spiker985, are you sure those were coming from `symbol-provider-ctags`? I ask because [we've got a `tags.scm` for Rust](https://github.com/pulsar-edit/pulsar/blob/master/packages/language-rust-bundled/grammars/tree-sitter-rust/queries/tags.scm?rgh-link-date=2024-01-06T19%3A31%3A08Z), so it would've been using that unless `symbol-provider-tree-sitter` had been disabled.
> 
> Just making sure we haven't got a bug we need to deal with at this late hour. If you're unsure which provider was supplying symbols, there's a setting to show the provider name in the list.

I toggled the setting for showing where suggestions come from, let me re-check

File: package.json
Comment (confused-Techie) [inline package.json:1]: We need to ensure we add `symbol-provider-ctags` and `symbol-provider-tree-sitter` to the `packageDependencies` section of the `package.json` here.

**Review by confused-Techie — CHANGES_REQUESTED** [review]
I've given the code a once over, and things seem pretty awesome! Just found one little oversight that I've added a request to change for, otherwise I think on the code only end we are good to go.

File: package.json
Comment (savetheclocktower) [reply package.json:1]: Does that just need to mirror what's present in `dependencies`? That's weird.

Comment (confused-Techie) [reply package.json:1]: Yeah it's a little odd. But if memory serves, that section has purposes two fold:
* It is what actually determines what's considered a core package to Pulsar
* Ensures these packages are part of the preload process

**Review by confused-Techie — APPROVED** [review]
Alright, I'm adding my approval to the code only portion of the PR, thanks for addressing my concerns so quickly.

And with Spikers awesome review of functionality, I think we are good to go, lets get this one merged!

**Pull request merged**